src/logger-mixin/mixins/logger-mixin.ts

```diff
@@ -0,0 +1,105 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: loopback-next-extension-starter
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+// tslint:disable:no-any
+
+import {Constructor} from '@loopback/context';
+
+/**
+* A mixin class for Application that creates a .logger()
+* function to register a Logger automatically. Also overrides
+* component function to allow it to register Logger's automatically.
+*
+* ```ts
+*
+* class MyApplication extends LoggerMixin(Application) {}
+* ```
+*/
+export function LoggerMixin<T extends Constructor<any>>(superClass: T) {
+  return class extends superClass {
+    // A mixin class has to take in a type any[] argument!
+    constructor(...args: any[]) {
+      super(...args);
+      if (!this.options) this.options = {};
+
+      if (this.options.loggers) {
+        for (const logger of this.options.loggers) {
+          this.logger(logger);
+        }
+      }
+
+      if (this.options.components) {
+        // Super would have already mounted the component
+        for (const component of this.options.components) {
+          this.component(component);
```

Comment:
`this.component` is already called in the base Application constructor, which means you should be able to do this:
```ts
if (this.options.components) {
  for (const component of this.options.components) {
    mountComponentLogger(component);
  }
}
```

Reply:
This is not needed at all. For the mixin to work correctly, it's enough to override `component` method (as is already done below). 

`Application`'s constructor, invoked via `super(...args)` at the top of this method, will call `this.component`, there is no need to call it explicitly again. Unless I am missing something about how the prototypal inheritance and `class` keyword works in JavaScript?

Reply:
Using type Application throws a whole series of errors such as using internal type `ApplicationOptions`, etc. I think it's because we don't have an interface for `Application`. Even setting it to `Constructor<{}> results in errors.